Optimizes Reftracking (Bigly) (Plus harddel fixes)

LemonInTheDark opened the pull request:
## About The Pull Request

### Reftracking BS

Alllright so reftracking is slow, really really slow.
That's a problem for me, both because I want it to be fast so I can more efficiently torture players by running it on live, but also because it impedes both local and CI runs. 

So I've set out to micro optimize the DoSearchVar proc, one of the hottest in the game.
I've done this in a few different ways.

#### The simple shit

Removing redundant proc args
Yeeting assoc arg setting (extra cost)
Moving if statements around to prioritize the more common case
Ignoring empty lists.

#### The not simple shit

Throwing our snowflake list checking into the sun
(Background, byond has some special lists that cannot be accessed like an assoc list, trying to will lead to runtimes)
The way we handle this involves inspecting their ref string, and it eats a LOT of time.

Faster then to mark all the lists we know are special by var name, and then use try/catch to detect and silence anything that sneaks through (this is on the order of like 1/3 per run, kinda curious what they are tbh) 
Thanks to MSO for the idea for this btw.

Removes the vars and logic that tied ref searching to clients. 
It's not how this code is used, and it slows everything else down for really no reason

Added support for handing in a known "hanging reference" count, and then searching for that. 
This lets us early exit the ref search if we find everything we were looking for, which is REALLY powerful, and why I asked for refcount() in the first place.

### Harddel Fixes

[Fixes some harddels w gulag stuff born of the 515 one way ref issues](https://github.com/tgstation/tgstation/commit/046d7daa03454fac501acb9a3960ea8306247da8)

[Ensures proximity cameras clean their ref to their proximity datum if it's deleted](https://github.com/tgstation/tgstation/commit/ff607e9ccb841d253c39502211ad9442a9cd362d)

[Deleting a pipe connected via the gas_machine_connector datum to a machine should also delete that machine (harddel fix)](https://github.com/tgstation/tgstation/commit/9eecca22e70b4a61ad48c04ddc9c726ef0c6713b)
## Why It's Good For The Game

All this combined speeds up refsearching massively, on the order of hundreds of seconds, and makes it far less time consuming for both CI and running on live. 
I'll be bullying some servers semi soon, want to see what I can cut out.

LemonInTheDark committed:
Fixes some harddels w gulag stuff born of the 515 one way ref issues

LemonInTheDark committed:
Ensures proximity cameras clean their ref to their proximity datum if it's deleted

LemonInTheDark committed:
Deleting a pipe connected via the gas_machine_connector datum to a machine should also delete that machine (harddel fix)

LemonInTheDark committed:
Speeds up reftracking significantly

Ok so I did a bunch of shit, some of it semi simple

Removed some args, removed some inefficent arg passing, moved shit
around to allow for the faster path, just don't handle empty lists.

Some not

The snowflake list check was eating a lot of time, so I've done a bunch
of bullshit for it
Rather then doing the weirdo list ref thing to figure out if it'll
error, we filter out the vars we know for sure will bite us in the ass
first, and then use try/catch to block errors and mark any offending
lists as such

Removes all the fluff that existed to tie running a ref search to
clients. It's fully automatic now, so there's no reason to have it (Also
none ever used it)

Added support for handing in a known "hanging reference" count, and then
searching for that. This lets us early exit the ref search if we find
everything we were looking for.

All this combined speeds up refsearching massively, on the order of
hundreds of seconds, and makes it far less time consuming for both CI
and running on live. Coming soon to a server near you

LemonInTheDark committed:
whoop

LemonInTheDark committed:
undef

LemonInTheDark committed:
Small micro, fixes tests

In unit tests we shouldn't do this check for every proc call, instead
just create the list when required

Also since we've introduced a ref count var, it needs to default to
INFINITY for ref victims so we can get to actually search

LemonInTheDark committed:
you saw nothin

LemonInTheDark committed:
LAZYACCESS

ZephyrTFA commented:
@LemonInTheDark not sure if you know this pr actually breaks ref tracking (according to CI)

LemonInTheDark commented:
yeah IK. S drafted cause I need to look into it, but it's really annoying without a debugger

LemonInTheDark committed:
fixes tests (adds logging for specials cause I'm curious

LemonInTheDark committed:
removes search_time

It isn't doing anything anymore since these are different instances of
these objects, and I don't want to confuse people

LemonInTheDark commented:
All works now, throwing er on sybil to test and also harvest hard deletes

Ghommie commented on code/modules/admin/view_variables/reference_tracking.dm:
```diff
@@ -90,73 +84,116 @@
+				// If we find all other refs before it I want to early exit, and if we don't I want to keep searching past it
+				var/ignore_ref = FALSE
+				var/list/queues = SSgarbage.queues
+				for(var/list/queue in queues)
```
Do you think we could use as anything here? There shouldn't be any issue if the `queue` is null.
```suggestion
				for(var/list/queue as anything in queues)
```

LemonInTheDark replied:
I uh, didn't realize that typechecked, tho it doesn't really matter in this case, will if you want

Ghommie reviewed (commented):
Skimming the logs, it seems that in many a case of harddel, a portion of the expected references is cleared before being found, making searches fruitless in several cases. This is often limited to datums which tend to constantly show up in the tracker, like chat messages. It works well for the rest.

LemonInTheDark commented:
> Skimming the logs, it seems that in many a case of harddel, a portion of the expected references is cleared before being found, making searches fruitless in several cases. This is often limited to datums which tend to constantly show up in the tracker, like chat messages. It works well for the rest.

This isn't an issue of refs being cleared before they're found, though that is possible. it's an issue of refs being stored in places we cannot access (in this case, https://www.byond.com/forum/post/2907707 is causing harddels by locking up a proc forever)
